File: fs/src/main/scala/farjs/fs/FSPluginUi.scala
package farjs.fs

import farjs.filelist.stack.WithStacks
import farjs.filelist.{FileListData, FileListPluginUiProps, FileListState}
import farjs.fs.FSPluginUi._
import farjs.fs.popups._
import scommons.react._

import scala.scalajs.js

object FSPluginUi {

  private[fs] var drive: UiComponent[DriveControllerProps] = DriveController
  private[fs] var foldersHistory: UiComponent[FoldersHistoryControllerProps] = FoldersHistoryController
  private[fs] var folderShortcuts: UiComponent[FolderShortcutsControllerProps] = FolderShortcutsController

  def unapply(arg: FSPluginUi): Option[(Option[Boolean], Boolean, Boolean)] = {
    Some((
      arg.showDrivePopupOnLeft,
      arg.showFoldersHistoryPopup,
      arg.showFolderShortcutsPopup
    ))
  }
}

class FSPluginUi(val showDrivePopupOnLeft: Option[Boolean] = None,
                 val showFoldersHistoryPopup: Boolean = false,
                 val showFolderShortcutsPopup: Boolean = false
                ) extends FunctionComponent[FileListPluginUiProps] {

  protected def render(compProps: Props): ReactElement = {
    val stacks = WithStacks.useStacks()
    val props = compProps.plain

    def onChangeDir(isLeft: Boolean): js.Function1[String, Unit] = { dir =>
      val currStack =
        if (isLeft) stacks.left.stack
        else stacks.right.stack

      if (currStack.peek() != currStack.peekLast()) {
        currStack.clear()
      }

      val stackItem = currStack.peekLast[FileListState]()
      stackItem.getData().foreach { case FileListData(dispatch, actions, state) =>
        if (dir != state.currDir.path) {
          dispatch(actions.changeDir(
            dispatch = dispatch,
            path = "",
            dir = dir
          ))
        }
      }
    }
    
    val onChangeDirInActivePanel: js.Function1[String, Unit] =
      onChangeDir(stacks.left.stack.isActive)

    <.>()(
      <(drive())(^.plain := DriveControllerProps(
        dispatch = props.dispatch,
        showDrivePopupOnLeft = showDrivePopupOnLeft match {
          case Some(v) => v
          case None => js.undefined
        },
        onChangeDir = { (dir, isLeft) =>
          onChangeDir(isLeft)(dir)
        },
        onClose = props.onClose
      ))(),
      
      <(foldersHistory())(^.plain := FoldersHistoryControllerProps(
        showPopup = showFoldersHistoryPopup,
        onChangeDir = onChangeDirInActivePanel,
        onClose = props.onClose
      ))(),
      
      <(folderShortcuts())(^.plain := FolderShortcutsControllerProps(
        showPopup = showFolderShortcutsPopup,
        onChangeDir = onChangeDirInActivePanel,
        onClose = props.onClose
      ))()
    )
  }
}

File: fs/FSPluginUi.mjs
/**
 * @import { FileListPluginUiProps } from "@farjs/filelist/FileListPlugin.mjs";
 */
import React from "react";
import WithStacks from "@farjs/filelist/stack/WithStacks.mjs";
import DriveController from "./popups/DriveController.mjs";
import FoldersHistoryController from "./popups/FoldersHistoryController.mjs";
import FolderShortcutsController from "./popups/FolderShortcutsController.mjs";

const h = React.createElement;

/**
 * @typedef {{
 *  readonly showDrivePopupOnLeft?: boolean;
 *  readonly showFoldersHistoryPopup: boolean;
 *  readonly showFolderShortcutsPopup: boolean;
 * }} FSPluginUiOptions
 */

/**
 * @param {FSPluginUiOptions} options
 */
function FSPluginUi(
  {
    showDrivePopupOnLeft,
    showFoldersHistoryPopup,
    showFolderShortcutsPopup,
  } = {
    showDrivePopupOnLeft: undefined,
    showFoldersHistoryPopup: false,
    showFolderShortcutsPopup: false,
  },
) {
  /**
   * @param {FileListPluginUiProps} props
   */
  const FSPluginUiComp = (props) => {
    const { drive, foldersHistory, folderShortcuts } = FSPluginUi;

    const stacks = WithStacks.useStacks();

    /** @type {(isLeft: boolean) => (dir: string) => void} */
    function onChangeDir(isLeft) {
      return (dir) => {
        const currStack = isLeft ? stacks.left.stack : stacks.right.stack;

        if (currStack.peek() !== currStack.peekLast()) {
          currStack.clear();
        }

        const stackItem = currStack.peekLast();
        const data = stackItem.getData();
        if (data) {
          const { dispatch, actions, state } = data;
          if (dir != state.currDir.path) {
            dispatch(actions.changeDir(dispatch, "", dir));
          }
        }
      };
    }

    const onChangeDirInActivePanel = onChangeDir(stacks.left.stack.isActive);

    return h(
      React.Fragment,
      null,

      h(drive, {
        dispatch: props.dispatch,
        showDrivePopupOnLeft,
        onChangeDir: (dir, isLeft) => {
          onChangeDir(isLeft)(dir);
        },
        onClose: props.onClose,
      }),

      h(foldersHistory, {
        showPopup: showFoldersHistoryPopup,
        onChangeDir: onChangeDirInActivePanel,
        onClose: props.onClose,
      }),

      h(folderShortcuts, {
        showPopup: showFolderShortcutsPopup,
        onChangeDir: onChangeDirInActivePanel,
        onClose: props.onClose,
      }),
    );
  };

  FSPluginUiComp.displayName = "FSPluginUi";

  return FSPluginUiComp;
}

FSPluginUi.drive = DriveController;
FSPluginUi.foldersHistory = FoldersHistoryController;
FSPluginUi.folderShortcuts = FolderShortcutsController;

export default FSPluginUi;
